control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,351 @@
+		return err
+	}
+
+	return nil
+}
+
+// getServiceResource converts the given Consul service and metadata as a Consul resource API record.
+func (r *Controller) getServiceResource(svc *pbcatalog.Service, name, namespace, partition string, meta map[string]string) *pbresource.Resource {
+	return &pbresource.Resource{
+		Id:       getServiceID(name, namespace, partition),
+		Data:     common.ToProtoAny(svc),
+		Metadata: meta,
+	}
+}
+
+func getServiceID(name, namespace, partition string) *pbresource.ID {
+	return &pbresource.ID{
+		Name: name,
+		Type: &pbresource.Type{
+			Group:        "catalog",
+			GroupVersion: "v1alpha1",
+			Kind:         "Service",
+		},
+		Tenancy: &pbresource.Tenancy{
+			Partition: partition,
+			Namespace: namespace,
+		},
+	}
+}
+
+// getServicePorts converts Kubernetes Service ports data into Consul service ports.
+func getServicePorts(service corev1.Service) []*pbcatalog.ServicePort {
+	ports := make([]*pbcatalog.ServicePort, 0, len(service.Spec.Ports)+1)
+
+	for _, p := range service.Spec.Ports {
+		ports = append(ports, &pbcatalog.ServicePort{
+			VirtualPort: uint32(p.Port),
+			TargetPort:  p.TargetPort.String(),
+			Protocol:    common.GetPortProtocol(p.AppProtocol),
+		})
+	}
+
+	// Append Consul service mesh port in addition to discovered ports.
+	//TODO(zalimeni) Omit if zero mesh ports present in service endpoints? Or could there be some
+	// use of mesh-inject at the service level in the webhook as well?
+	ports = append(ports, &pbcatalog.ServicePort{
+		TargetPort: "mesh",
+		Protocol:   pbcatalog.Protocol_PROTOCOL_MESH,
+	})
+
+	return ports
+}
+
+// getServiceVIPs returns the VIPs to associate with the registered Consul service. This will contain the Kubernetes
+// Service ClusterIP if it exists.
+//
+// Note that we always provide this data regardless of whether TProxy is enabled, deferring to individual proxy configs
+// to decide whether it's used.
+func (r *Controller) getServiceVIPs(service corev1.Service) []string {
+	if parsedIP := net.ParseIP(service.Spec.ClusterIP); parsedIP == nil {
+		r.Log.Info("skipping service registration virtual IP assignment due to invalid or unset ClusterIP", "name", service.Name, "ns", service.Namespace, "ip", service.Spec.ClusterIP)
+		return nil
+	}
+	return []string{service.Spec.ClusterIP}
+}
+
+func getServiceMeta(service corev1.Service) map[string]string {
+	meta := map[string]string{
+		constants.MetaKeyKubeNS: service.Namespace,
+		metaKeyManagedBy:        constants.ManagedByEndpointsValue,
+	}
+	//TODO(zalimeni) This is adapted from v1 pod-based meta injection - verify correct and needed
+	for k, v := range service.Annotations {
+		if strings.HasPrefix(k, constants.AnnotationMeta) && strings.TrimPrefix(k, constants.AnnotationMeta) != "" {
+			if v == "$SERVICE_NAME" {
+				meta[strings.TrimPrefix(k, constants.AnnotationMeta)] = service.Name
+			} else {
+				meta[strings.TrimPrefix(k, constants.AnnotationMeta)] = v
+			}
+		}
+	}
```

Comment:
I wasn't sure whether we wanted to keep allowing for custom meta via annotations, but seemed feasible to adapt from the v1 controller. Thoughts? Happy to cut if we're not sure about supporting it, or the interpolation of `$SERVICE_NAME`.

Reply:
I left this out for now, but I don't have a strong opinion.

Reply:
Omitted for brevity 🚀 TODO'd for follow-up